### PR TITLE
select filled elements by clicking inside

### LINKED ISSUE #104: Select filled elements by clicking inside
Figma allows to select filled elements by clicking inside the element
Screen Recording 2020-01-02 at 23 19 20

### PR DESCRIPTION
**Pull request by dwelle**
fixes #104 

Alright, this one is crazy, but it kinda works. I have no idea what I'm doing exactly around the ellipse, so don't ask me.

Also, the algo for `diamond` shape could in theory be factored out and used for any rectangle, but I kept a simpler one for regular `rectangle` due to perf.

Related, there's an issue how we handle coordinates when we flip shapes horizontally/vertically. I hacked-normalized it when detecting hitbox for diamond shape, but in the long run we need to normalize globally, because this results in several other bugs. Will create an issue for this, soon.

Also, for perf, I'd probably make `backgroundColor` `null` to indicate no bg is present, instead of `"transparent"`, so that we don't need to compare strings.

Also note, that when we implement #15, all the code in `collison.ts` will likely break.

excalidraw_filled_hitbox

### CODE REVIEW

File: src/element/collision.ts [hunk 1]
```diff
@@ -78,6 +93,42 @@ export function hitTest(
       leftY
     ] = getDiamondPoints(element);
 
+    if (element.backgroundColor !== "transparent") {
+      // TODO: remove this when we normalize coordinates globally
+      if (topY > bottomY) [bottomY, topY] = [topY, bottomY];
+      if (rightX < leftX) [leftX, rightX] = [rightX, leftX];
```
Comment (dwelle) [inline src/element/collision.ts]: This hack is necessary when we flip the shape (via resize) horizontally/vertically.

Comment (voluntadpear) [reply src/element/collision.ts]: Perhaps #342 helps with this

Comment (dwelle) [reply src/element/collision.ts]: Oh, cool, didn't notice your PR. Yes, it'll likely fix that. So maybe wait for your PR to land before merging this?

**Comment by vjeux** [comment]
The math looks reasonable and it seems to be working. Shipit! Thanks :)

**Pull request merged**